File: Decorator/Java/src/condiments/Soy.java
package eu.svetlin.design.patterns.decorator.condiments;

import eu.svetlin.design.patterns.decorator.beverages.Beverage;

public class Soy extends CondimentDecorator {

    public Soy(Beverage beverage){
        this.beverage = beverage;
    }

    private double getSoyCost() {
        switch (beverage.getSize()) {
            case GRANDE:
                return .20;
            case VENTI:
                return .25;
            case TALL:
            default:
                return .15;
        }
    }
    @Override
    public double cost() {
        return beverage.cost() + getSoyCost();
    }

    @Override
    public String getDescription() {
        return beverage.getDescription() + ", Soy";
    }
}

File: Decorator/TypeScript/src/condiments/Soy.ts
import Beverage from '../beverages/Beverage';
import CondimentDecorator from './CondimentDecorator';

class Soy extends CondimentDecorator {
  constructor(beverage: Beverage) {
    super();
    this.beverage = beverage;
  }
  getDescription() {
    return this.beverage.getDescription() + ', Soy';
  }

  cost(): number {
    return this.beverage.cost() + 0.15;
  }
}

export default Soy;
